dbms/src/Storages/DeltaMerge/LocalIndexerScheduler.cpp

```diff
@@ -0,0 +1,415 @@
+// limitations under the License.
+
+#include <Common/Exception.h>
+#include <Storages/DeltaMerge/LocalIndexerScheduler.h>
+#include <common/logger_useful.h>
+
+namespace DB::DM
+{
+
+bool operator==(const LocalIndexerScheduler::FileID & lhs, const LocalIndexerScheduler::FileID & rhs)
+{
+    if (lhs.index() != rhs.index())
+        return false;
+
+    auto index = lhs.index();
+    if (index == 0)
+    {
+        return std::get<LocalIndexerScheduler::DMFileID>(lhs).id == std::get<LocalIndexerScheduler::DMFileID>(rhs).id;
+    }
+    else if (index == 1)
+    {
+        return std::get<LocalIndexerScheduler::ColumnFileTinyID>(lhs).id
+            == std::get<LocalIndexerScheduler::ColumnFileTinyID>(rhs).id;
+    }
+    return false;
+}
+
+LocalIndexerScheduler::LocalIndexerScheduler(const Options & options)
+    : logger(Logger::get())
+    , pool(std::make_unique<ThreadPool>(options.pool_size, options.pool_size, options.pool_size + 1))
+    , pool_max_memory_limit(options.memory_limit)
+    , pool_current_memory(0)
+{
+    // QueueSize = PoolSize+1, because our scheduler will try to schedule next task
+    // right before the current task is finished.
+
+    LOG_INFO(
+        logger,
+        "Initialized LocalIndexerScheduler, pool_size={}, memory_limit_mb={:.1f}",
+        options.pool_size,
+        static_cast<double>(options.memory_limit) / 1024 / 1024);
+
+    if (options.auto_start)
+        start();
+}
+
+LocalIndexerScheduler::~LocalIndexerScheduler()
+{
+    LOG_INFO(logger, "LocalIndexerScheduler is destroying. Waiting scheduler and tasks to finish...");
+
+    // First quit the scheduler. Don't schedule more tasks.
+    is_shutting_down = true;
+    {
+        std::unique_lock lock(mutex);
+        scheduler_need_wakeup = true;
+        scheduler_notifier.notify_all();
+    }
+
+    if (is_started)
+        scheduler_thread.join();
+
+    // Then wait all running tasks to finish.
+    pool.reset();
+
+    LOG_INFO(logger, "LocalIndexerScheduler is destroyed");
+}
+
+void LocalIndexerScheduler::start()
+{
+    if (is_started)
+        return;
+
+    scheduler_thread = std::thread([this]() { schedulerLoop(); });
+    is_started = true;
+}
+
+void LocalIndexerScheduler::waitForFinish()
+{
+    while (true)
+    {
+        std::unique_lock lock(mutex);
+        if (all_tasks_count == 0 && running_tasks_count == 0)
+            return;
+        on_finish_notifier.wait(lock);
+    }
+}
+
+void LocalIndexerScheduler::pushTask(const Task & task)
+{
+    if (pool_max_memory_limit > 0 && task.request_memory > pool_max_memory_limit)
+        throw Exception(fmt::format(
+            "Requests memory exceeds limit (request={} limit={})",
+            task.request_memory,
+            pool_max_memory_limit));
+
+    std::unique_lock lock(mutex);
+
+    const auto internal_task = std::make_shared<InternalTask>(InternalTask{
+        .user_task = task,
+        .created_at = Stopwatch(),
+        .scheduled_at = Stopwatch{}, // Not scheduled
```

Comment:
```suggestion
        .created_at = Stopwatch(),
        .scheduled_at = Stopwatch(), // Not scheduled
```
should init by the same form?